pyscord/core/http.py

```diff
@@ -0,0 +1,145 @@
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+import asyncio
+from enum import Enum
+from typing import Dict, Any, Protocol
+
+from aiohttp import ClientSession
+from aiohttp.client import _RequestContextManager
+from aiohttp.typedefs import StrOrURL
+
+from ..exceptions import NotFoundError, BadRequestError, NotModifiedError, \
+    UnauthorizedError, ForbiddenError, MethodNotAllowedError, RateLimitError, \
+    ServerError, HTTPError
+
+
+class RequestMethod(Enum):
+    GET = 0
+    POST = 1
+    DELETE = 2
+    PUT = 3
+    OPTIONS = 4
+
+
+class HttpCallable(Protocol):
+    def __call__(self, url: StrOrURL, *, allow_redirects: bool = True,
+                 **kwargs: Any) -> "_RequestContextManager":
+        ...
+
+
+class HTTPClient:
+    def __init__(self, token: str, version: int = 9, ttl: int = 4):
+        """
+        Instantiate a new HttpApi object.
+
+        :param token: Discord API token
+
+        Keyword Arguments:
+        :param version: The discord API version.
+                        See https://discord.com/developers/docs/reference#api-versioning.
+        :param ttl: Max amount of attempts after
+                            error code 5xx
+        """
+        self.header: Dict[str, str] = {
+            "Authorization": f"Bot {token}"
+        }
+        self.endpoint: str = f"https://discord.com/api/v{version}"
+        self.max_ttl: int = ttl
+        self.__http_exceptions: Dict[int, HTTPError] = {
+            304: NotModifiedError(),
+            400: BadRequestError(),
+            401: UnauthorizedError(),
+            403: ForbiddenError(),
+            404: NotFoundError(),
+            405: MethodNotAllowedError(),
+            439: RateLimitError()
+        }
+
+    async def __send(self, method: RequestMethod, route: str, *,
+                     __ttl: int = None):
+        # TODO: Fix docs
+        # TODO: Implement logging
+        __ttl = self.max_ttl if __ttl is None else __ttl
+
+        if __ttl == 0:
+            raise ServerError(f"Maximum amount of retries for `{route}`.")
+
+        async with ClientSession() as session:
+            methods: Dict[RequestMethod, HttpCallable] = {
+                RequestMethod.GET: session.get,
+                RequestMethod.POST: session.post,
+                RequestMethod.DELETE: session.delete,
+                RequestMethod.PUT: session.put,
+                RequestMethod.OPTIONS: session.options
+            }
+
+            sender = methods.get(method)
+
+            if not sender:
+                raise RuntimeError("Invalid RequestMethod has been passed.")
+
+            async with sender(f"{self.endpoint}/{route}",
+                              headers=self.header) as res:
+
+                if res.status in [200, 201, 204]:
```

Comment:
Maybe `res.status >= 200 and res.status < 300`?

Reply:
Response object does offer a `ok` property which is defined by the following:
```py
    @property
    def ok(self):
        """Returns True if :attr:`status_code` is less than 400, False if not.

        This attribute checks if the status code of the response is between
        400 and 600 to see if there was a client error or a server error. If
        the status code is between 200 and 400, this will return True. This
        is **not** a check to see if the response code is ``200 OK``.
        """
        try:
            self.raise_for_status()
        except HTTPError:
            return False
        return True
```

Which seems to be ok for what we doing since the status code can be grouped as:
- `100–199` Informational responses
- `200–299` Successful responses
- `300–399` Redirects
- `400–499` Client errors
- `500–599` Server errors
(*more @:* https://developer.mozilla.org/en-US/docs/Web/HTTP/Status)

Reply:
Did this